lib/src/rules/eol_last.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2016, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Put an end of line for non-empty files.';
+
+const _details = r'''
+
+**DO** put an end of line for non-empty files.
+
+**BAD:**
+```dart
+a {
+}
+```
+
+**GOOD:**
+```dart
+b {
+}
+
+''';
+
+class EolLast extends LintRule implements NodeLintRule {
+  EolLast()
+      : super(
+            name: 'eol_last',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    var endToken = node.endToken;
+    var previous = endToken.previous;
+    if (previous == null) {
+      return;
+    }
+    var comments = endToken.precedingComments;
+    if (previous.end == endToken.offset ||
+        (comments != null && comments.end == endToken.offset)) {
```

Comment:
The `preceedingComments` is a (doubly) linked list of tokens with one token per comment, ordered such that the token with the smallest offset is first in the list. If there's more than one comment at the end of the file then this will look at the end of the first comment when it should be looking at the end of the last comment. For example:
```
class C {}
// This is just a placeholder.
// This file should be replaced with useful code.
```

An easier solution might be to pass the `LinterContext` to the `_Visitor` constructor and store it in an instance field, then use `context.currentUnit.content` to get the content of the file and check to see whether it ends with an end-of-line marker (assuming that it isn't an empty string).

Reply:
Thanks a lot for the hint.